File: 2_ReactWeb/src/services/llm/tokenEstimationSettings.ts
import { fetchJson } from "../http/httpClient";

export type TokenEstimationSettings = {
  ascii_chars_per_token: number;
  other_chars_per_token: number;
  message_overhead_tokens: number;
  image_placeholder_tokens: number;
};

export type TokenEstimationSettingsResponse = {
  settings: TokenEstimationSettings;
  default_settings: TokenEstimationSettings;
  updated_at: string | null;
};

export function getTokenEstimationSettings() {
  return fetchJson<TokenEstimationSettingsResponse>(
    "/api/llm/token-estimation-settings",
  );
}

export function saveTokenEstimationSettings(settings: TokenEstimationSettings) {
  return fetchJson<TokenEstimationSettingsResponse>(
    "/api/llm/token-estimation-settings",
    {
      body: JSON.stringify({ settings }),
      method: "PUT",
    },
  );
}

export function estimateJsonTokens(value: unknown) {
  return fetchJson<{ token_count: number }>(
    "/api/llm/token-estimation-settings/estimate-json",
    {
      body: JSON.stringify({ value }),
      method: "POST",
    },
  );
}

File: 1_PythonServer/app/domain/llm/token_estimation_settings.py
from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class TokenEstimationSettings:
    ascii_chars_per_token: float
    other_chars_per_token: float
    message_overhead_tokens: int
    image_placeholder_tokens: int
    updated_at: str | None = None


DEFAULT_TOKEN_ESTIMATION_SETTINGS = TokenEstimationSettings(
    ascii_chars_per_token=4.0,
    other_chars_per_token=2.0,
    message_overhead_tokens=4,
    image_placeholder_tokens=2000,
)
